### PR TITLE
move file when drop in blank space of explorer

### PR DESCRIPTION
**Pull request by kpge**
fix: #5947
Signed-off-by: kpge <gekangping@126.com>

#### What it does
drap file to root
As there is no clear conclusion of `What are expectations in the multi-root scenario`, the expectations in the multi-root workspace is the same with vscode:
drap file in multi-root project

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
I wonder is there a way to avoid shaking effect while d&d.

File: packages/filesystem/src/browser/file-tree/file-tree-widget.tsx [hunk 1]
```diff
@@ -175,6 +175,18 @@ export class FileTreeWidget extends TreeWidget {
         }
     }
 
+    protected getTargetDir(node: TreeNode | undefined): DirNode | undefined {
```
Comment (akosyakov) [inline packages/filesystem/src/browser/file-tree/file-tree-widget.tsx]: `getDropTargetDirNode`

Comment (kpge) [reply packages/filesystem/src/browser/file-tree/file-tree-widget.tsx]: ok

**Comment by kpge** [comment]
> I wonder is there a way to avoid shaking effect while d&d.

The shaking effect is caused by here, when move from one node to another, the `unstyle` handler will be called:
https://github.com/theia-ide/theia/blob/0eebcf368461d4e138befb4d07448c07f06942de/packages/core/src/browser/view-container.ts#L804-L805

**Comment by kpge** [comment]
> I wonder is there a way to avoid shaking effect while d&d.

when move one node to another node, this `unstyle` handler will be called:
https://github.com/theia-ide/theia/blob/0eebcf368461d4e138befb4d07448c07f06942de/packages/core/src/browser/view-container.ts#L804-L805
I don't kown why this event happened.

**Comment by akosyakov** [comment]
> I don't kown why this event happened.

View containers can have multiple views. And then one can drag and drop them. We need to check btw that it still works well.

One can do it by installing https://github.com/microsoft/vscode-extension-samples/tree/master/tree-view-sample from sources and changing its package.json that views are added to the explorer view container.

**Comment by kpge** [comment]
> I don't kown why this event happened.

`shaking effect` may be is a bug of 'dragleave' : https://stackoverflow.com/questions/10867506/dragleave-of-parent-element-fires-when-dragging-over-children-elements
dragleave in the childElement will trigger the parent's `dragleave` event.

**Comment by lmcbout** [comment]
Tested on a single root workspace .
1- D&D to another folder: OK when the drop location folder is **visible**
2- D&D to root  if whole root is visible: OK
3- D&D to root when scroll is needed to reach the bottom area before Dropping the file: **NOT OK,** the auto-scroll is not active when dragging is over the last item of the tree, so unable to reach the root location while dragging a file.

DnDRootToBig

**Comment by vince-fugnitto** [comment]
@lmcbout 

> 3- D&D to root when scroll is needed to reach the bottom area before Dropping the file: **NOT OK,** the auto-scroll is not active when dragging is over the last item of the tree, so unable to reach the root location while dragging a file.

Please see the following issue https://github.com/theia-ide/theia/issues/2918 and PR https://github.com/theia-ide/theia/pull/4726.
I don't believe this PR is addressing the problem already in master.

**Comment by akosyakov** [comment]
Agree with @vince-fugnitto, scroll never worked, it should be handled separately.

**Comment by akosyakov** [comment]
@kpge let's go with shaking effect for now, but please open a follow up issue and with label `help wanted`

**Comment by lmcbout** [comment]
Testing this PR with single root and seems there is a refresh missing when we Drag a file to the root folder( The file is removed from sub-folder, but doesn't show in the root folder until a browser refreshed) . If I drag the file to a sub-folder, then it seems fine. See Video.
DnDRefresh

**Comment by kpge** [comment]
@lmcbout The `auto-refresh ` in root folder can't be useful may be the issue #938.

**Comment by kpge** [comment]
> @kpge let's go with shaking effect for now, but please open a follow up issue and with label `help wanted`

I have opened a new issue to follow up the shaking effect, is this PR can be merged?

**Comment by akosyakov** [comment]
@kpge i've tested now again and shaking effect is pretty annoying when you want to drop into the another folder. We have to sort it out before merging :( VS Code does not shake for example.

**Comment by akosyakov** [comment]
> D&D a file to the first root folder: same issue, need a browser refresh to see the file in the root folder. To test it, I need to D&D a file over the root folder itself. (Tested with Chrome on Ubuntu )

Could you share a workspace to which one can reproduce it easier? I am not sure that it is related to this PR, since file refresh happens on fs events, so seems to be nsfw does not deliver again. Maybe you try to reproduce it before view containers and it was the same then file a bug.

**Comment by lmcbout** [comment]
Just created a small workspace
Tested on Chrome and Firefox
OS: Ubuntu 16.06
Theia: Latest from this morning + patch from @kpge 

[testFolder.gz](https://github.com/theia-ide/theia/files/3525117/testFolder.gz)
To extract the file folder to your directory:  tar -xvf

Test:
Select a file  (.theia/jacques.test) and Drag  it to the root folder

**Comment by kpge** [comment]
when I add a timer in `dragleave` handler, the shaking effect is gone:
drap file
is this change can been push to this PR?
```diff
diff --git a/packages/core/src/browser/view-container.ts b/packages/core/src/browser/view-container.ts
index f0aba85b0..f5d4f5f56 100644
--- a/packages/core/src/browser/view-container.ts
+++ b/packages/core/src/browser/view-container.ts
@@ -759,8 +759,12 @@ export class ViewContainerPart extends BaseWidget {
 
     protected registerDND(): Disposable {
         this.header.draggable = true;
+        let dropZoneTimer: NodeJS.Timer;
         const style = (event: DragEvent) => {
             event.preventDefault();
+            if (dropZoneTimer) {
+                clearTimeout(dropZoneTimer);
+            }
             const part = ViewContainerPart.closestPart(event.target);
             if (part instanceof HTMLElement) {
                 if (this.canBeDropTarget) {
@@ -771,9 +775,15 @@ export class ViewContainerPart extends BaseWidget {
         const unstyle = (event: DragEvent) => {
             event.preventDefault();
             const part = ViewContainerPart.closestPart(event.target);
-            if (part instanceof HTMLElement) {
-                part.classList.remove('drop-target');
+            if (dropZoneTimer) {
+                clearTimeout(dropZoneTimer);
             }
+            dropZoneTimer = setTimeout(() => {
+                if (part instanceof HTMLElement) {
+                    part.classList.remove('drop-target');
+                }
+            }, 100);
+
         };
         return new DisposableCollection(
             addEventListener(this.header, 'dragstart', event => {
```

**Comment by akosyakov** [comment]
@kpge we should change how drag and drop works in view containers. A part cannot be target to itself. This blueish outline should not appear at all if one does not pull another part into it.

**Comment by kpge** [comment]
> Select a file (.theia/jacques.test) and Drag it to the root folder

@lmcbout 
thanks. I 'v reproduced it, and I'm looking at it now

**Comment by akosyakov** [comment]
@kpge I know how to fix flickering/shaking effect. I will do another PR on top of yours.

**Comment by akosyakov** [comment]
@kpge I've opened a PR against your branch to fix shaking: https://github.com/kpge/theia/pull/4

**Comment by kpge** [comment]
@lmcbout 
I've fixed the bug when d&d in the root folder, please test it :)

**Pull request merged**